src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1502,3 +1530,227 @@ RelationUsesHeapAccessMethodOrNone(Relation relation)
 	return true;
 #endif
 }
+
+
+/*
+ * UndistributeTable undistributes the given table. The undistribution is done by
+ * creating a new table, moving everything to the new table and dropping the old one.
+ * So the oid of the table is not preserved.
+ *
+ * The undistributed table will have the same name, columns and rows. It will also have
+ * partitions, views, sequences of the old table. Finally it will have everything created
+ * by GetTableConstructionCommands function, which include indexes. These will be
+ * re-created during undistribution, so their oids are not preserved either (except for
+ * sequences). However, their names are preserved.
+ *
+ * The tables with references are not supported. The function gives an error if there are
+ * any references to or from the table.
+ *
+ * The dropping of old table is done with CASCADE. Anything not mentioned here will
+ * be dropped.
+ */
+void
+UndistributeTable(Oid relationId)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table"),
+						errdetail("No such distributed table exists. "
+								  "Might have already been undistributed.")));
+	}
+
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table."),
+						errdetail("The table is not distributed.")));
+	}
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+
+	List *tableBuildingCommands = GetTableBuildingCommands(relationId, true);
+	List *tableConstructionCommands = GetTableConstructionCommands(relationId);
+
+	tableConstructionCommands = list_concat(tableConstructionCommands,
+											GetViewCreationCommandsOfTable(relationId));
+
+	int spiResult = SPI_connect();
+	if (spiResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager")));
+	}
+
+	char *relationName = get_rel_name(relationId);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	ereport(NOTICE, (errmsg("Starting undistribution of %s.%s",
+							schemaName, relationName)));
+
+	if (PartitionedTable(relationId))
+	{
+		ereport(NOTICE, (errmsg("Undistributing the partitions of %s.%s",
+								schemaName, relationName)));
```

Comment:
I think it's a bit nicer to use quote_qualified_identifier(schemaName, relationName)